File: 202410/239.sliding-window-maximum.cpp
/*
 * @lc app=leetcode id=239 lang=cpp
 * @lcpr version=20000
 *
 * [239] Sliding Window Maximum
 */

// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start

// 使用单调队列
class MonotonicQueue
{
private:
    deque<int> data;

public:
    void push(int val)
    {
        while (!data.empty() && data.back() < val)
        {
            data.pop_back();
        }
        data.push_back(val);
    }
    void pop(int cur)
    {
        if (!data.empty() && data.front() == cur)
        {
            data.pop_front();
        }
    }
    int max()
    {
        return data.front();
    }
};

class Solution
{
public:
    vector<int>
    maxSlidingWindow(vector<int> &nums, int k)
    {
        MonotonicQueue window;
        vector<int> res;

        for (int i = 0; i < nums.size(); i++)
        {
            if (i < k - 1)
            {
                window.push(nums[i]);
            }
            else
            {
                window.push(nums[i]);
                res.push_back(window.max());
                window.pop(nums[i - k + 1]);
            }
        }

        return res;
    }
};
// @lc code=end

/*
// @lcpr case=start
// [1,3,-1,-3,5,3,6,7]\n3\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

 */

File: 202503/239.Sliding Window Maximum.java
/*
 * @lc app=leetcode id=239 lang=java
 * @lcpr version=30100
 *
 * [239] Sliding Window Maximum
 */

// @lc code=start

import java.util.LinkedList;
import java.util.Queue;
import java.util.Deque;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        MonoQueue mq = new MonoQueue();
        int n = nums.length;
        int[] res = new int[n - k + 1];

        for (int i = 0; i < n; i++) {
            mq.add(nums[i]);
            if (mq.getSize() < k) {
                continue;
            }
            if (mq.getSize() > k) {
                mq.remove();
            }
            res[i - k + 1] = mq.getMax();
        }

        return res;
    }
}

class MonoQueue {
    Queue<Integer> q = new LinkedList<Integer>();
    Deque<Integer> maxQ = new LinkedList<Integer>();

    public void add(int val) {
        q.add(val);
        while (!maxQ.isEmpty() && maxQ.getLast() < val) {
            maxQ.removeLast();
        }
        maxQ.addLast(val);
    }

    public void remove() {
        int first = q.remove();
        if (first == maxQ.getFirst()) {
            maxQ.removeFirst();
        }
    }

    public int getMax() {
        return maxQ.getFirst();
    }

    public int getSize() {
        return q.size();
    }

}
// @lc code=end



/*
// @lcpr case=start
// [1,3,-1,-3,5,3,6,7]\n3\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

 */

